sklearn/utils/__init__.py

```diff
@@ -186,7 +186,9 @@ def _array_indexing(array, key, key_dtype, axis):
 
 def _pandas_indexing(X, key, key_dtype, axis):
     """Index a pandas dataframe or a series."""
-    if hasattr(key, "shape"):
+    if hasattr(key, "shape") and not np.isscalar(key):
```

Comment:
Could we use `_is_arraylike_not_scalar` instead?
https://github.com/scikit-learn/scikit-learn/blob/74ddf019860705417b89d09a30048be2df7fd1cf/sklearn/utils/validation.py#L257-L264

Reply:
At first we could not because a tuple is an array-like and there's a different logic for the tuple case right after. But it turns out that we can get rid of the pandas < 0.25 workaround and that there's no reason to make it a list when it's a tuple and an array when it's not. We can always make it an array.

So in the end I think we can use `_is_array_like_not_scalar` and simplify the whole logic. Done in [a642ec9](https://github.com/scikit-learn/scikit-learn/pull/24932/commits/a642ec97a67931b643d28e2d3d5b0ae17cef46f5)